deployment/postgres-init-db/sql/functions/hasura/plan_migration_functions.sql

```diff
@@ -0,0 +1,199 @@
+create table hasura.migrate_plan_to_model_return_value(result text);
+/*
+* This function does the following:
+*     * creates a snapshot of the specified plan
+*     * updates the specified plan to have model_id = _new_model_id
+*     * invalidates the activity validations, which will trigger the activity validator to run again
+* It will not update the plan if:
+*     * user has incorrect permissions (see default_user_roles for details)
+*     * there are open merge requests for the given plan
+*     * the given plan or model does not exist
+*/
+create function hasura.migrate_plan_to_model(_plan_id integer, _new_model_id integer, hasura_session json)
+  returns hasura.migrate_plan_to_model_return_value
+  volatile
+  language plpgsql as $$
+declare
+  _requester_username  text;
+  _function_permission permissions.permission;
+  _old_model_id        integer;
+begin
+  _requester_username := (hasura_session ->> 'x-hasura-user-id');
+  _function_permission := permissions.get_function_permissions('migrate_plan_to_model', hasura_session);
+  perform permissions.raise_if_plan_merge_permission('migrate_plan_to_model', _function_permission);
+  if not _function_permission = 'NO_CHECK' then
+    call permissions.check_general_permissions('migrate_plan_to_model', _function_permission, _plan_id,
+                                               _requester_username);
+  end if;
+
+  if not exists(select id from merlin.plan where id = _plan_id) then
+    raise exception 'Plan % does not exist, not proceeding with plan migration.', _plan_id;
+  end if;
+
+  if not exists(select id from merlin.mission_model where id = _new_model_id) then
+    raise exception 'Model % does not exist, not proceeding with plan migration.', _new_model_id;
+  end if;
+
+
+  -- Check for open merge requests
+  if exists(select
+            from merlin.merge_request mr
+            where mr.plan_id_receiving_changes = _plan_id
+              and status in ('pending', 'in-progress')) then
+    raise exception 'Cannot migrate plan %: it has open merge requests.', _plan_id;
+  end if;
+
+  -- Get the old model ID associated with the plan
+  select model_id into _old_model_id from merlin.plan where id = _plan_id;
+
+  -- Create snapshot before migration
+  perform merlin.create_snapshot(_plan_id,
+                                 'Migration from model ' || _old_model_id ||
+                                 ' to model ' || _new_model_id ||
+                                 ' on ' || NOW(),
+                                 'Automatic snapshot before attempting migration from model id ' || _old_model_id ||
+                                 ' to model id ' || _new_model_id ||
+                                 ' on ' || NOW(),
+                                 _requester_username);
```

Comment:
Nonblocking, ignorable: Condensing snapshot title and removing incorrect phrase "attempting":
```suggestion
  perform merlin.create_snapshot(_plan_id,
                                 'Migration from model ' || _old_model_id ||
                                 ' to model ' || _new_model_id,
                                 'Automatic snapshot before migrating from model id ' || _old_model_id ||
                                 ' to model id ' || _new_model_id ||
                                 ' on ' || NOW(),
                                 _requester_username);
```

Reply:
Updated in ac9c10f7cc1b2300d1a6d8cb39fc85e9b8ca3705